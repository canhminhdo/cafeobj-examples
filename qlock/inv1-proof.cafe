in qlock.cafe

--> base case
open QLOCK .
    op p : -> Pid .
    red inv1(init, p) .
close

--> induction case for want
--> case 1.1
open QLOCK .
    op s : -> Sys .
    ops p r : -> Pid .
    eq pc(s, r) = rs .
    eq p = r .
    red inv1(s, p) implies inv1(want(s, r), p) .
close

--> case 1.2.1.1
open QLOCK .
    op s : -> Sys .
    ops p r : -> Pid .
    eq pc(s, r) = rs .
    eq (p = r) = false .
    eq pc(s, p) = cs .
    eq top(queue(s)) = p .
    red inv1(s, p) implies inv1(want(s, r), p) .
close

--> case 1.2.1.2
open QLOCK .
    op s : -> Sys .
    ops p r : -> Pid .
    eq pc(s, r) = rs .
    eq (p = r) = false .
    eq pc(s, p) = cs .
    eq (top(queue(s)) = p) = false .
    red inv1(s, p) implies inv1(want(s, r), p) .
close

--> case 1.2.2
open QLOCK .
    op s : -> Sys .
    ops p r : -> Pid .
    eq pc(s, r) = rs .
    eq (p = r) = false .
    eq (pc(s, p) = cs) = false .
    red inv1(s, p) implies inv1(want(s, r), p) .
close

--> case 2
open QLOCK .
    op s : -> Sys .
    ops p r : -> Pid .
    eq (pc(s, r) = rs) = false .
    red inv1(s, p) implies inv1(want(s, r), p) .
close