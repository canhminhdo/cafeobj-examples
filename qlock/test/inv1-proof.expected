--> base case
(true):Bool
--> induction case for want
--> case 1.1
(true):Bool
--> case 1.2.1.1
(true):Bool
--> case 1.2.1.2
(true):Bool
--> case 1.2.2
(true):Bool
--> case 2
(true):Bool
